c/certificateService.c

```diff
@@ -0,0 +1,188 @@
+
+/*
+  This program and the accompanying materials are
+  made available under the terms of the Eclipse Public License v2.0 which accompanies
+  this distribution, and is available at https://www.eclipse.org/legal/epl-v20.html
+  
+  SPDX-License-Identifier: EPL-2.0
+  
+  Copyright Contributors to the Zowe Project.
+*/
+
+#ifdef METTLE 
+#include <metal/metal.h>
+#include <metal/stddef.h>
+#include <metal/stdio.h>
+#include <metal/stdlib.h>
+#include <metal/string.h>
+#include "metalio.h"
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#endif
+
+#include "httpserver.h"
+#include "json.h"
+#include "http.h"
+
+#pragma linkage(IRRSIM00, OS)
+
+#define MAP_CERTIFICATE_TO_USERNAME 0x0006
+#define SUCCESS_RC 0
+#define SAF_FAILURE_RC 8
+#define RACF_FAILURE_RC 8
+
+// Reason codes representing potential errors within R_Usermap macro
+#define PARAMETER_LIST_ERROR_RC 4
+#define NO_MAPPING_TO_USERID_RC 16
+#define NOT_AUTHORIZED_RC 20
+#define CERTIFICATE_NOT_VALID_RC 28
+#define NOTRUST_CERTIFICATE_RC 32
+#define NO_IDENTITY_FILTER_RC 48
+
+typedef _Packed struct _RUsermapParamList { 
+   double workarea[128]; 
+    int safRcAlet, returnCode;
+    int racfRcAlet, returnCodeRacf;
+    int racfReasonAlet, reasonCodeRacf;
+    int fcAlet;
+    short functionCode;
+    int  optionWord;
+    char useridLengthRacf; 
+    char useridRacf[8];
+    int certificateLength;
+    char *certificate;
+    short applicationIdLength;
+    char applicationId[246];
+    short distinguishedNameLength;
+    char distinguishedName[246];
+    short registryNameLength;
+    char registryName[255];
+    
+} RUsermapParamList;
+
+static void setValidResponseCode(HttpResponse *response, int rc, int returnCode, int returnCodeRacf, int reasonCodeRacf) {
+  if (rc == SUCCESS_RC && returnCode == SUCCESS_RC && returnCodeRacf == SUCCESS_RC && reasonCodeRacf == SUCCESS_RC) {
+    setResponseStatus(response, 200, "OK");
+    return;
+  } else if(rc != SUCCESS_RC) {
+    if(returnCode == SAF_FAILURE_RC && returnCodeRacf == RACF_FAILURE_RC) {
+      if(reasonCodeRacf == PARAMETER_LIST_ERROR_RC) {
+        setResponseStatus(response, 400, "Bad request");
+        return;
+      } else if (
+                 reasonCodeRacf == NO_MAPPING_TO_USERID_RC || 
+                 reasonCodeRacf == NOT_AUTHORIZED_RC || 
+                 reasonCodeRacf == CERTIFICATE_NOT_VALID_RC || 
+                 reasonCodeRacf == NOTRUST_CERTIFICATE_RC || 
+                 reasonCodeRacf == NO_IDENTITY_FILTER_RC
+                 ) {
+        setResponseStatus(response, 401, "Unauthorized");
+        return;
+      }
+    } 
+  } 
+
+  setResponseStatus(response, 500, "Internal server error");
+  return;
+}
+
+static void handleInvalidMethod(HttpResponse *response) {
+    jsonPrinter *p = respondWithJsonPrinter(response);
+      
+    setResponseStatus(response, 405, "Method Not Allowed");
+    setDefaultJSONRESTHeaders(response);
+    addStringHeader(response, "Allow", "GET");
```

Comment:
If the goal of this method is to inform callers that only POST requests are acceptable, then the value of the Allow header should be POST.  I would also suggest renaming the method "respond..." (e.g. respondAllowPOST) rather than "handle..." because handle... naming is typically used for top-level web service functions.  It also makes clear in calling code that no *other* method in the execution flow should attempt to respond to the client, either before or after the call to this method.